# Data Analysis on eBay Kleinanzeigen Used Car Ads

We will be working with a dataset that comes from a German classified ads section of eBay called Kleinanzeigen that sells used cars.

The original dataset was scraped and placed out to [Kaggle](https://www.kaggle.com/orgesleka/used-cars-database/data). For our purposes, we are going to be using the dataset that was prepared by the DataQuest team that has been narrowed down to 50,000 rows and purposely made with data issues for data cleaning.

The data dictionary for this dataset is as follows:

- dateCrawled - When this ad was first crawled. All field-values are taken from this date.
- name - Name of the car.
- seller - Whether the seller is private or a dealer.
- offerType - The type of listing
- price - The price on the ad to sell the car.
- abtest - Whether the listing is included in an A/B test.
- vehicleType - The vehicle Type.
- yearOfRegistration - The year in which the car was first registered.
- gearbox - The transmission type.
- powerPS - The power of the car in PS.
- model - The car model name.
- kilometer - How many kilometers the car has driven.
- monthOfRegistration - The month in which the car was first registered.
- fuelType - What type of fuel the car uses.
- brand - The brand of the car.
- notRepairedDamage - If the car has a damage which is not yet repaired.
- dateCreated - The date on which the eBay listing was created.
- nrOfPictures - The number of pictures in the ad.
- postalCode - The postal code for the location of the vehicle.
- lastSeenOnline - When the crawler saw this ad last online.

The goal of this project is to clean this dataset and use it for analysis.

In [1]:
import pandas as pd
import numpy  as np

autos = pd.read_csv("autos.csv",encoding = "Windows-1252")

In [2]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine_

In [3]:
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

From what we can see in this dataset, there are 20 columns with a total of 50,000 rows. There are some nulls in the vehicleType, gearbox, model, fuelType, and notRepairedDamage columns, but none of these are greater than 20% of the total dataset. Also, there are date values that are stored as strings.

We will start to clean this dataset in preparation for analysis.

In [4]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

We are going to make some changes here by placing underscores to separate out words within a column and to make columns more indicative of the data contained within the column.

In [6]:
autos.columns = ['date_crawled','name','seller','offer_type','price','abtest','vehicle_type','registration_year'
                 ,'gear_box','power_ps','model','odometer','registration_month','fuel_type','brand','unrepaired_damage'
                 ,'ad_created','num_of_pictures','postal_code','last_seen']
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

    gear_box  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  num_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00                0   
1         bmw              nein  2016-04-04 00:00:00                0   
2  volkswagen              nein  2016-03-26 00:00:00                0   
3       smart              nein  2016-03-12 00:00:00                0   
4        ford              nein  2016-04-01 00:00:00                0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

## Initial Data Exploration and Cleaning

We will now take a look at the data to see if there are columns that need to be cleaned up and what data needs to be investigated further.

In [8]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-29 23:42:13  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year gear_box      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  num_of_pictures   postal_code  \
count                 50000          50000.0  50000.000000   
unique                   76              NaN           NaN   
top     2016-04-03 00:00:00              NaN           NaN   
freq                   1946              NaN           NaN   
mean                    NaN              0.0  50813.627300   
std                     NaN              0.0  25779.747957   
min                     NaN              0.0   1067.000000   
25%                     NaN              0.0  30451.000000   
50%                     NaN              0.0  49577.000000   
75%                     NaN              0.0  71540.000000   
max                     NaN              0.0  99998.000000   

                  last_seen  
count                 50000  
unique                39481  
top     2016-04-07 06:17:27  
freq                      8  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

As we can see in this description of the dataset, that nearly all of the values in the seller and offer_type columns are the same. 

The num_of_photos column may need to be updated to numeric values since all of the data is reporting as NaN. The price and odometer columns also need to have special characters removed and converted over to numeric.

We will take a look at the odometer column first to see if there are any anonmalies that we need to account for when cleaning the data.

In [12]:
autos_od = autos["odometer"].value_counts()
print(autos_od)

150,000km    32424
125,000km     5170
100,000km     2169
90,000km      1757
80,000km      1436
70,000km      1230
60,000km      1164
50,000km      1027
5,000km        967
40,000km       819
30,000km       789
20,000km       784
10,000km       264
Name: odometer, dtype: int64


The odometer column shows that we need to remove commas and km before converting these values to numeric. We will also convert the column header name to odometer_km.

In [20]:
autos["odometer"] = autos["odometer"].replace(",","").replace("km","").astype(int)
autos.rename(columns = {"odometer" : "odometer_km"}, inplace=True)
autos_od = autos["odometer_km"].value_counts()
print(autos_od)

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: odometer_km, dtype: int64


We will repeat this process for the price column.

In [14]:
autos_pr = autos["price"].value_counts()
print(autos_pr)

$0         1421
$500        781
$1,500      734
$2,500      643
$1,200      639
$1,000      639
$600        531
$3,500      498
$800        498
$2,000      460
$999        434
$750        433
$900        420
$650        419
$850        410
$700        395
$4,500      394
$300        384
$2,200      382
$950        379
$1,100      376
$1,300      371
$3,000      365
$550        356
$1,800      355
$5,500      340
$350        335
$1,250      335
$1,600      327
$1,999      322
           ... 
$83,000       1
$17,149       1
$68,500       1
$55,996       1
$10,988       1
$1,189        1
$8,099        1
$19,780       1
$3,710        1
$9,270        1
$23,790       1
$5,895        1
$55,800       1
$6,998        1
$32,150       1
$1,770        1
$5,924        1
$29           1
$255          1
$23,799       1
$4,222        1
$3,120        1
$1,930        1
$1,070        1
$3,996        1
$37,480       1
$15,823       1
$1,906        1
$554          1
$46,990       1
Name: price, Length: 235

The price column only needs to have commas and dollar signs removed before being converted to float.

In [22]:
autos["price"] = autos["price"].replace(",","").replace("$","").astype(float)
autos_pr = autos["price"].value_counts()
print(autos_pr)

0.0           1421
500.0          781
1500.0         734
2500.0         643
1200.0         639
1000.0         639
600.0          531
800.0          498
3500.0         498
2000.0         460
999.0          434
750.0          433
900.0          420
650.0          419
850.0          410
700.0          395
4500.0         394
300.0          384
2200.0         382
950.0          379
1100.0         376
1300.0         371
3000.0         365
550.0          356
1800.0         355
5500.0         340
1250.0         335
350.0          335
1600.0         327
1999.0         322
              ... 
2225.0           1
69997.0          1
139997.0         1
69999.0          1
4780.0           1
8930.0           1
21599.0          1
15911.0          1
10000000.0       1
5180.0           1
919.0            1
1247.0           1
5998.0           1
27020.0          1
21888.0          1
46500.0          1
2001.0           1
2459.0           1
345000.0         1
34940.0          1
2785.0           1
5248.0      

Now, we will take a look at the NaN values in the num_of_pictures column.

In [27]:
auto_photos = autos["num_of_pictures"].value_counts()
print(auto_photos)

0    50000
Name: num_of_pictures, dtype: int64


Since there are no significant values to be drawn from this column, we can go ahead and remove it and the seller and offer type columns from the dataset.

In [28]:
autos = autos.drop(["num_of_pictures","offer_type","seller"],axis=1)

## Exploring the Odometer and Prices

As we have cleaned up the data for the odometer and the price columns, we can take a closer look at them. We will start by looking at the odometer_km column.

In [34]:
print(autos["odometer_km"].value_counts())

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: odometer_km, dtype: int64


The values in the odometer_km column are all rounded values, so this can indicate that whem posters are placing their ad on the website that they can only select rounded values instead of putting specific data values. The amount of higher mileage vehicles compared to lower mileage vehicles is higher.

In [40]:
print(autos["price"].unique().shape[0])
print(autos["price"].describe())
print(autos["price"].value_counts().sort_index(ascending=False))

2357
count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64
99999999.0       1
27322222.0       1
12345678.0       3
11111111.0       2
10000000.0       1
3890000.0        1
1300000.0        1
1234566.0        1
999999.0         2
999990.0         1
350000.0         1
345000.0         1
299000.0         1
295000.0         1
265000.0         1
259000.0         1
250000.0         1
220000.0         1
198000.0         1
197000.0         1
194000.0         1
190000.0         1
180000.0         1
175000.0         1
169999.0         1
169000.0         1
163991.0         1
163500.0         1
155000.0         1
151990.0         1
              ... 
66.0             1
65.0             5
60.0             9
59.0             1
55.0             2
50.0            49
49.0             4
47.0             1
45.0             4
40.0             6
35.0        

The price column also seems rounded since there are no cents listed within the prices. We can also see that there are 1421 listings that are $0, which is 3 percent of the dataset and these rows could possibly be removed.

In [45]:
print(autos["price"].value_counts().sort_index(ascending=False).head(20))

99999999.0    1
27322222.0    1
12345678.0    3
11111111.0    2
10000000.0    1
3890000.0     1
1300000.0     1
1234566.0     1
999999.0      2
999990.0      1
350000.0      1
345000.0      1
299000.0      1
295000.0      1
265000.0      1
259000.0      1
250000.0      1
220000.0      1
198000.0      1
197000.0      1
Name: price, dtype: int64


There are also really high priced listings that are odd such as 12345678, 11111111, and 99999999, which may have been a user entering a value just to have this required field populated. 

In [44]:
print(autos["price"].value_counts().sort_index(ascending=True).head(20))

0.0     1421
1.0      156
2.0        3
3.0        1
5.0        2
8.0        1
9.0        1
10.0       7
11.0       2
12.0       3
13.0       2
14.0       1
15.0       2
17.0       3
18.0       1
20.0       4
25.0       5
29.0       1
30.0       7
35.0       1
Name: price, dtype: int64


On the lower end of the prices, there are listings that are below 30. Since eBay is an auctioning site, posters can list items for as low as a 1.

To filter this dataset to more realistic values, we can focus on those listings from 1 to 350,000 since the values that are above that seem to be placeholder values that are put in by users just to fill in a required field.

In [48]:
autos = autos[autos["price"].between(1,350000)]
print(autos["price"].describe())

count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price, dtype: float64


## Exploring the date columns

There are the following columns in the dataset:

- date_crawled
- last_seen
- ad_created
- registration_month
- registration_year

The date_crawled, last_seen, and ad_created columns were all converted into strings upon import since the values in these columns are dates and times.

In [49]:
print(autos[["date_crawled","last_seen","ad_created"]].head(5))

          date_crawled            last_seen           ad_created
0  2016-03-26 17:47:46  2016-04-06 06:45:54  2016-03-26 00:00:00
1  2016-04-04 13:38:56  2016-04-06 14:45:08  2016-04-04 00:00:00
2  2016-03-26 18:57:24  2016-04-06 20:15:37  2016-03-26 00:00:00
3  2016-03-12 16:58:10  2016-03-15 03:16:28  2016-03-12 00:00:00
4  2016-04-01 14:38:50  2016-04-01 14:38:50  2016-04-01 00:00:00


We will take a look at the distributions of each of these columns with just focusing on the dates.

In [55]:
autos["date_crawled"].str[:10].value_counts(normalize=True,dropna=False).sort_index()

2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
2016-03-25    0.031607
2016-03-26    0.032204
2016-03-27    0.031092
2016-03-28    0.034860
2016-03-29    0.034099
2016-03-30    0.033687
2016-03-31    0.031834
2016-04-01    0.033687
2016-04-02    0.035478
2016-04-03    0.038608
2016-04-04    0.036487
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64

From what we can see here, most of the data was crawled between 03/05/2016-04/07/2016.

In [56]:
autos["last_seen"].str[:10].value_counts(normalize=True,dropna=False).sort_index()

2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-17    0.028086
2016-03-18    0.007351
2016-03-19    0.015834
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-22    0.021373
2016-03-23    0.018532
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-26    0.016802
2016-03-27    0.015649
2016-03-28    0.020859
2016-03-29    0.022341
2016-03-30    0.024771
2016-03-31    0.023783
2016-04-01    0.022794
2016-04-02    0.024915
2016-04-03    0.025203
2016-04-04    0.024483
2016-04-05    0.124761
2016-04-06    0.221806
2016-04-07    0.131947
Name: last_seen, dtype: float64

This also seems to be the case with the dates in the last_seen column that the dates are between 03/05/2016-04/07/2016. This value indicates the date that the crawler last saw the listing if the car was sold or the listing pulled by the poster, and there is a higher percentage of values on the last three dates of the distribution. This may only mean that this was the last time that the crawler saw the listing and not a spike in sales or ads being removed.

In [57]:
autos["ad_created"].str[:10].value_counts(normalize=True,dropna=False).sort_index()

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
2015-12-30    0.000021
2016-01-03    0.000021
2016-01-07    0.000021
2016-01-10    0.000041
2016-01-13    0.000021
2016-01-14    0.000021
2016-01-16    0.000021
2016-01-22    0.000021
2016-01-27    0.000062
2016-01-29    0.000021
2016-02-01    0.000021
2016-02-02    0.000041
2016-02-05    0.000041
2016-02-07    0.000021
2016-02-08    0.000021
2016-02-09    0.000021
2016-02-11    0.000021
2016-02-12    0.000041
2016-02-14    0.000041
2016-02-16    0.000021
2016-02-17    0.000021
2016-02-18    0.000041
2016-02-19    0.000062
2016-02-20    0.000041
2016-02-21    0.000062
                ...   
2016-03-09    0.033151
2016-03-10    0.031895
2016-03-11    0.032904
2016-03-12    0.036755
2016-03-13    0.017008
2016-03-14    0.035190
2016-03-15    0.034016
2016-03-16    0.030125
2016-03-17    0.031278
2016-03-18    0.013590
2016-03-19    0.033687
2016-03-20    0.037949
2016-03-21 

There is quite a distribution in date ranges with the majority falling during the timeframe that the data was scraped, but there are outliers that were present for months before the data was scraped.

In [58]:
autos["registration_year"].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

For the registration year, there appears to be some outliers that may have the same issue as with the price where a user just put a place holder value in such as the minimum value being 1000 and the maximum value being 9999.

## Dealing with Incorrect Values in the Registration Year

Since there are values in the registration year that were just entered in as place holder values, we will need to filter these values out since they are not valid.

In [60]:
autos["registration_year"].value_counts().sort_index(ascending=False)

9999       3
9000       1
8888       1
6200       1
5911       1
5000       4
4800       1
4500       1
4100       1
2800       1
2019       2
2018     470
2017    1392
2016    1220
2015     392
2014     663
2013     803
2012    1310
2011    1623
2010    1589
2009    2085
2008    2215
2007    2277
2006    2670
2005    2936
2004    2703
2003    2699
2002    2486
2001    2636
2000    3156
        ... 
1964      12
1963       8
1962       4
1961       6
1960      23
1959       6
1958       4
1957       2
1956       4
1955       2
1954       2
1953       1
1952       1
1951       2
1950       3
1948       1
1943       1
1941       2
1939       1
1938       1
1937       4
1934       2
1931       1
1929       1
1927       1
1910       5
1800       2
1111       1
1001       1
1000       1
Name: registration_year, Length: 95, dtype: int64

Since the data set was scraped between March 2016 and April 2016, we can drop any registrations that are greater than 2016. 

After some research through [Wikipedia](https://en.wikipedia.org/wiki/Car) and the [Daimler](https://www.daimler.com/company/tradition/company-history/1886-1920.html) websites, the car was first mass produced in Germany in 1902. We can then narrow down the dataset to those rows that have a registration year between 1900 and 2016.

In [67]:
autos = autos[autos["registration_year"].between(1900,2016)]
autos["registration_year"].describe()

count    46681.000000
mean      2002.910756
std          7.185103
min       1910.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       2016.000000
Name: registration_year, dtype: float64

We can see that the majority of the car registration years are within 20 years of when the data was scraped.

## Exploring Price by Brand

Now that we have a cleaner dataset, we can take a look to explore the brands of cars in the listings and their prices.

In [69]:
autos["brand"].value_counts(normalize=True)

volkswagen        0.211264
bmw               0.110045
opel              0.107581
mercedes_benz     0.096463
audi              0.086566
ford              0.069900
renault           0.047150
peugeot           0.029841
fiat              0.025642
seat              0.018273
skoda             0.016409
nissan            0.015274
mazda             0.015188
smart             0.014160
citroen           0.014010
toyota            0.012703
hyundai           0.010025
sonstige_autos    0.009811
volvo             0.009147
mini              0.008762
mitsubishi        0.008226
honda             0.007840
kia               0.007069
alfa_romeo        0.006641
porsche           0.006127
suzuki            0.005934
chevrolet         0.005698
chrysler          0.003513
dacia             0.002635
daihatsu          0.002506
jeep              0.002271
subaru            0.002142
land_rover        0.002099
saab              0.001649
jaguar            0.001564
daewoo            0.001500
trabant           0.001392
r

We can see that the majority of the brands that were listed on the eBay website are German brands Volkswagen, BMW, Opel, Mercedes Benz, and Audi. Since a lot of the other brands have less than a 5% representation in the dataset, we can go ahead and only focus on the more popular brands.

In [80]:
brand_counts = autos["brand"].value_counts(normalize = True)
most_common_brands = brand_counts[brand_counts > .05].index
print(most_common_brands)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')


In [83]:
top_5_pct_brands = {}

for cb in most_common_brands:
    selected_rows = autos[autos["brand"] == cb]
    mean_rows = selected_rows["price"].mean()
    top_5_pct_brands[cb] = mean_rows 
    
top_5_pct_brands

{'volkswagen': 5402.410261610221,
 'bmw': 8332.820517811953,
 'opel': 2975.2419354838707,
 'mercedes_benz': 8628.450366422385,
 'audi': 9336.687453600594,
 'ford': 3749.4695065890287}

From what we can see, the most expensive brands are Audi, Mercedes Benz, and BMW. Ford and Open are the least expensive with Volkswagen being the in between brand. This could explain why this is the most popular brand listed in our analysis.

## Exploring Mileage Data for Most Popular Brands

We will go ahead and take a look at average mileage and see if there is a relation to the mean price.

In [89]:
top_brands_series = pd.Series(top_5_pct_brands)
top_brands_mean_price = pd.DataFrame(top_brands_series, columns = ["Mean Price"])
top_brands_mean_price

Mean Price
volkswagen     5402.410262
bmw            8332.820518
opel           2975.241935
mercedes_benz  8628.450366
audi           9336.687454
ford           3749.469507

In [88]:
top_5_pct_mileage = {}

for row in most_common_brands:
    selected_rows = autos[autos["brand"] == row]
    mean_mileage = selected_rows["odometer_km"].mean()
    top_5_pct_mileage[row] = mean_mileage
    
top_5_pct_mileage

{'volkswagen': 128707.15879132022,
 'bmw': 132572.51313996495,
 'opel': 129310.0358422939,
 'mercedes_benz': 130788.36331334666,
 'audi': 129157.38678544914,
 'ford': 124266.01287159056}

In [93]:
top_brands_mileage_series = pd.Series(top_5_pct_mileage)
top_brands_mean_mileage = pd.DataFrame(top_brands_mileage_series, columns = ["Mean Mileage"])
top_brands_mean_mileage = top_brands_mean_mileage.sort_values("Mean Mileage",ascending = False)
top_brands_mean_mileage

Mean Mileage
bmw            132572.513140
mercedes_benz  130788.363313
opel           129310.035842
audi           129157.386785
volkswagen     128707.158791
ford           124266.012872

In [94]:
top_brands_mean_mileage["Mean Price"] = top_brands_mean_price
top_brands_mean_mileage

Mean Mileage   Mean Price
bmw            132572.513140  8332.820518
mercedes_benz  130788.363313  8628.450366
opel           129310.035842  2975.241935
audi           129157.386785  9336.687454
volkswagen     128707.158791  5402.410262
ford           124266.012872  3749.469507

In conclusion, the mean mileage between the most popular brands does not vary as much as the prices do. The more expensive brands though do have higher mean mileage as opposed to the lesser expensive brands.